In [2]:
import numpy as np
import pandas as pd

from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.datasets import load_boston
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

In [9]:
boston = load_boston()
X = boston.data
Y = boston.target
print("X shape is:", X.shape)
print("Y shape is:", Y.shape)

X shape is: (506, 13)
Y shape is: (506,)


### Function for model creation

In [15]:
def baseline_model():
#     Create model
    model = Sequential()
    model.add(Dense(13,
                    input_dim=13,
                    kernel_initializer="normal",
                    activation="relu"))
    model.add(Dense(1,kernel_initializer="normal"))
#     Compile model
    model.compile(loss="mse", optimizer="adam")
    return model

### Evaluate model

In [16]:
# Fix random seed 
seed = 7
np.random.seed(seed)
# Evaluate model with standardized dataset
estimator = KerasRegressor(build_fn=baseline_model,
                           epochs=100,
                           batch_size=5,
                           verbose=0)

In [18]:
kfold = KFold(n_splits=10, random_state=seed)
results = cross_val_score(estimator, X, Y, cv=kfold, verbose=1)
print("Results: %.2f (%.2f) MSE" % (results.mean(), results.std()))

Results: -35.38 (22.28) MSE


[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:  1.4min finished


### Evaluate model with standardized data

In [21]:
np.random.seed(seed)
estimators = []
estimators.append(("standardize", StandardScaler()))
estimators.append(("mlp", KerasRegressor(build_fn=baseline_model,
                                         epochs=50,
                                         batch_size=5,
                                         verbose=1)))
pipeline = Pipeline(estimators)
kfold = KFold(n_splits=10, random_state=seed)
results = cross_val_score(pipeline, X, Y, cv=kfold)
print("Standardized: %.2f (%.2f) MSE" % (results.mean(), results.std()))

Epoch 1/50
455/455 [==============================] - 1s 1ms/step - loss: 598.4831
Epoch 2/50
455/455 [==============================] - 0s 211us/step - loss: 553.8410
Epoch 3/50
455/455 [==============================] - 0s 259us/step - loss: 461.9125
Epoch 4/50
455/455 [==============================] - 0s 240us/step - loss: 338.5039
Epoch 5/50
455/455 [==============================] - 0s 236us/step - loss: 224.5539
Epoch 6/50
455/455 [==============================] - 0s 207us/step - loss: 142.7442
Epoch 7/50
455/455 [==============================] - 0s 220us/step - loss: 95.6675
Epoch 8/50
455/455 [==============================] - 0s 219us/step - loss: 70.4836
Epoch 9/50
455/455 [==============================] - 0s 217us/step - loss: 55.5772
Epoch 10/50
455/455 [==============================] - 0s 256us/step - loss: 46.0963
Epoch 11/50
455/455 [==============================] - 0s 216us/step - loss: 39.7335
Epoch 12/50
455/455 [==============================] - 0s 219us/step -

455/455 [==============================] - 0s 271us/step - loss: 16.7877
Epoch 47/50
455/455 [==============================] - 0s 257us/step - loss: 16.5079
Epoch 48/50
455/455 [==============================] - 0s 258us/step - loss: 16.3310
Epoch 49/50
455/455 [==============================] - 0s 266us/step - loss: 16.0245
Epoch 50/50
51/51 [==============================] - 0s 4ms/step
Epoch 1/50
455/455 [==============================] - 1s 2ms/step - loss: 616.1651
Epoch 2/50
455/455 [==============================] - 0s 306us/step - loss: 587.7353
Epoch 3/50
455/455 [==============================] - 0s 302us/step - loss: 500.9172
Epoch 4/50
455/455 [==============================] - 0s 248us/step - loss: 363.3188
Epoch 5/50
455/455 [==============================] - 0s 238us/step - loss: 230.7152
Epoch 6/50
455/455 [==============================] - 0s 224us/step - loss: 139.2870
Epoch 7/50
455/455 [==============================] - 0s 232us/step - loss: 91.6931
Epoch 8/50
455/

455/455 [==============================] - 0s 291us/step - loss: 16.6822
Epoch 42/50
455/455 [==============================] - 0s 259us/step - loss: 16.4184
Epoch 43/50
455/455 [==============================] - 0s 278us/step - loss: 16.2644
Epoch 44/50
455/455 [==============================] - 0s 287us/step - loss: 15.9331
Epoch 45/50
455/455 [==============================] - 0s 265us/step - loss: 15.6944
Epoch 46/50
455/455 [==============================] - 0s 248us/step - loss: 15.4516
Epoch 47/50
455/455 [==============================] - 0s 242us/step - loss: 15.2328
Epoch 48/50
455/455 [==============================] - 0s 277us/step - loss: 14.9512
Epoch 49/50
455/455 [==============================] - 0s 286us/step - loss: 14.7231
Epoch 50/50
51/51 [==============================] - 0s 4ms/step
Epoch 1/50
455/455 [==============================] - 1s 2ms/step - loss: 555.6708
Epoch 2/50
455/455 [==============================] - 0s 269us/step - loss: 511.3111
Epoch 3/50
455

455/455 [==============================] - 0s 283us/step - loss: 19.7599
Epoch 37/50
455/455 [==============================] - 0s 284us/step - loss: 19.4377
Epoch 38/50
455/455 [==============================] - 0s 296us/step - loss: 19.1784
Epoch 39/50
455/455 [==============================] - 0s 278us/step - loss: 18.7554
Epoch 40/50
455/455 [==============================] - 0s 287us/step - loss: 18.4199
Epoch 41/50
455/455 [==============================] - 0s 275us/step - loss: 18.1444
Epoch 42/50
455/455 [==============================] - 0s 295us/step - loss: 17.8220
Epoch 43/50
455/455 [==============================] - 0s 276us/step - loss: 17.5231
Epoch 44/50
455/455 [==============================] - 0s 280us/step - loss: 17.2904
Epoch 45/50
455/455 [==============================] - 0s 287us/step - loss: 17.0359
Epoch 46/50
455/455 [==============================] - 0s 277us/step - loss: 16.7624
Epoch 47/50
455/455 [==============================] - 0s 270us/step - loss: 

456/456 [==============================] - 0s 300us/step - loss: 17.5187
Epoch 32/50
456/456 [==============================] - 0s 293us/step - loss: 17.2306
Epoch 33/50
456/456 [==============================] - 0s 302us/step - loss: 16.9209
Epoch 34/50
456/456 [==============================] - 0s 306us/step - loss: 16.6818
Epoch 35/50
456/456 [==============================] - 0s 312us/step - loss: 16.4086
Epoch 36/50
456/456 [==============================] - 0s 299us/step - loss: 16.1410
Epoch 37/50
456/456 [==============================] - 0s 275us/step - loss: 15.8531
Epoch 38/50
456/456 [==============================] - 0s 282us/step - loss: 15.5879
Epoch 39/50
456/456 [==============================] - 0s 289us/step - loss: 15.3311
Epoch 40/50
456/456 [==============================] - 0s 302us/step - loss: 15.0396
Epoch 41/50
456/456 [==============================] - 0s 312us/step - loss: 14.7392
Epoch 42/50
456/456 [==============================] - 0s 321us/step - loss: 

456/456 [==============================] - 0s 271us/step - loss: 27.1699
Epoch 27/50
456/456 [==============================] - 0s 266us/step - loss: 26.7524
Epoch 28/50
456/456 [==============================] - 0s 300us/step - loss: 26.3577
Epoch 29/50
456/456 [==============================] - 0s 270us/step - loss: 25.9494
Epoch 30/50
456/456 [==============================] - 0s 266us/step - loss: 25.6312
Epoch 31/50
456/456 [==============================] - 0s 302us/step - loss: 25.2265
Epoch 32/50
456/456 [==============================] - 0s 270us/step - loss: 24.9395
Epoch 33/50
456/456 [==============================] - 0s 278us/step - loss: 24.6090
Epoch 34/50
456/456 [==============================] - 0s 254us/step - loss: 24.2631
Epoch 35/50
456/456 [==============================] - 0s 249us/step - loss: 23.8875
Epoch 36/50
456/456 [==============================] - 0s 271us/step - loss: 23.6035
Epoch 37/50
456/456 [==============================] - 0s 264us/step - loss: 

### Evaluate some other models to see what works best
#### Try a larger model

In [22]:
def larger_model():
#     Create model
    model = Sequential()
    model.add(Dense(13,
                    input_dim=13,
                    kernel_initializer="normal",
                    activation="relu"))
    model.add(Dense(6,
                    kernel_initializer="normal",
                    activation="relu"))
    model.add(Dense(1, kernel_initializer="normal"))
#     Compile model
    model.compile(loss="mse", optimizer="adam")
    return model

In [24]:
np.random.seed(seed)
estimators = []
estimators.append(("standardize", StandardScaler()))
estimators.append(("mlp", KerasRegressor(build_fn=larger_model,
                                         epochs=50,
                                         batch_size=5,
                                         verbose=0)))
pipeline = Pipeline(estimators)
kfold = KFold(n_splits=10, random_state=seed)
results = cross_val_score(pipeline, X, Y, cv=kfold)
print("Larger: %.2f (%.2f) MSE" % (results.mean(), results.std()))

Larger: -23.06 (26.46) MSE


#### Try a wider model 

In [25]:
def wider_model():
#     Create model
    model = Sequential()
    model.add(Dense(20,
                    input_dim=13,
                    kernel_initializer="normal",
                    activation="relu"))
    model.add(Dense(1, kernel_initializer="normal"))
#     Compile model
    model.compile(loss="mse", optimizer="adam")
    return model

In [26]:
np.random.seed(seed)
estimators = []
estimators.append(("standardize", StandardScaler()))
estimators.append(("mlp", KerasRegressor(build_fn=wider_model,
                                         epochs=100,
                                         batch_size=5,
                                         verbose=0)))
pipeline = Pipeline(estimators)
kfold = KFold(n_splits=10, random_state=seed)
results = cross_val_score(pipeline, X, Y, cv=kfold)
print("Wider: %.2f (%.2f) MSE" % (results.mean(), results.std()))

Wider: -22.86 (29.09) MSE


### The results show that a wider network outperforms a deeper network for this problem
#### MSE:
#### Baseline model: 29.53
#### Larger model: 23.06
#### Wider model: 22.86
### This shows that it is hard to predict what work